tests/test_quad/test_bayesquad/test_bq.py

```diff
@@ -219,3 +219,131 @@ def test_zero_function_gives_zero_variance_with_mle(rng):
     )
     assert bq_integral1.var == 0.0
     assert bq_integral2.var == 0.0
+
+
+def test_multilevel_bayesquad_from_data_input_handling(kernel, measure, rng):
+    """Test that inputs to multilevel BQ are handled properly."""
+    n_level = 3
+    fun_diff_evals_1 = ()
+    ns_1 = (3, 7, 2)
+    for l in range(n_level):
+        fun_diff_evals_1 += (np.zeros((ns_1[l],)),)
+    # No kernels given so defaults to ExpQuad
+    nodes_full = ()
+    for l in range(n_level):
+        nodes_full += (measure.sample(n_sample=ns_1[l], rng=rng),)
+    F, infos = multilevel_bayesquad_from_data(
+        nodes=nodes_full,
+        fun_diff_evals=fun_diff_evals_1,
+        measure=measure,
+    )
+    assert isinstance(F, Normal)
+    assert len(infos) == n_level
+    # Only one set of nodes
+    kernels_1 = n_level * (kernel,)
+    ns_2 = (7, 7, 7)
+    fun_diff_evals_2 = n_level * (np.zeros((ns_2[0],)),)
+    kernels_full = n_level * (kernel,)
+    nodes_1 = (measure.sample(n_sample=ns_2[0], rng=rng),)
+    F, infos = multilevel_bayesquad_from_data(
+        nodes=nodes_1,
+        fun_diff_evals=fun_diff_evals_2,
+        kernels=kernels_full,
+        measure=measure,
+    )
+    assert isinstance(F, Normal)
+    assert len(infos) == n_level
+    # Only one kernel and one set of nodes
+    F, infos = multilevel_bayesquad_from_data(
+        nodes=nodes_1,
+        fun_diff_evals=fun_diff_evals_2,
+        kernels=kernels_1,
+        measure=measure,
+    )
+    assert isinstance(F, Normal)
+    assert len(infos) == n_level
+    # Wrong number inputs should throw error
+    kernels_2 = (kernel, kernel)
+    with pytest.raises(ValueError):
+        _, _ = multilevel_bayesquad_from_data(
+            nodes=nodes_1,
+            fun_diff_evals=fun_diff_evals_2,
+            kernels=kernels_2,
+            measure=measure,
+        )
+    nodes_2 = (nodes_full[0], nodes_full[1])
+    with pytest.raises(ValueError):
+        _, _ = multilevel_bayesquad_from_data(
+            nodes=nodes_2,
+            fun_diff_evals=fun_diff_evals_2,
+            kernels=kernels_2,
+            measure=measure,
+        )
```

Comment:
can you factor out the exception raises tests into a separate test such that one test function checks the output type and shape, and the other the exception raises based on wrong input?